### PR TITLE
9562 New app launcher component

### PR DESCRIPTION
**Pull request by amaroks**
**Resolves [#9562](https://chartiq.kanbanize.com/ctrl_dashboard//cards/9562/details)**

-- This is in progress --

### CODE REVIEW

File: src-built-in/components/appLauncher2/src/components/AppDefinition.jsx [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+	render() {
+		const app = this.props.app
+		return (
+			<div className="app-item">
+				<div className="app-item-title">{app.friendlyName}</div>
+					<div className="app-item-tags">
+						{
+							app.tags.map((tag, index) => {
+								return <span key={index}>{tag}, </span>
```
Comment (carterbs) [inline src-built-in/components/appLauncher2/src/components/AppDefinition.jsx]: add a check here. `if(index !== app.tags.length)`, then render the `,`.

File: src-built-in/components/appLauncher2/src/components/Content.jsx [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+		super(props)
+		this.state = {
+			filter: {
+				text: ''
+			}
+		}
+	}
+
+	filterApps() {
+		const apps = this.props.activeFolder.appDefinitions
```
Comment (carterbs) [inline src-built-in/components/appLauncher2/src/components/Content.jsx]: Once you get to making this filter work, you'll want your filterSort component to set a value on the store; this component will listen for changes on that field. When the value changes, this component renders with the newly filtered list.

File: src-built-in/components/appLauncher2/src/components/LeftNav.jsx [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import React from  'react'
+import FoldersList from  './FoldersList'
+
+const bottomEntries = [
+	'New folder',
+	'New dashboard',
+	'App catalogue'
```
Comment (carterbs) [inline src-built-in/components/appLauncher2/src/components/LeftNav.jsx]: Go ahead and use american english here. "App Catalog"

File: src-built-in/components/appLauncher2/src/components/LeftNav.jsx [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+				  			className={className} key={index}>
+				  				{folder.name}
+				  			</div>
+				  			)
+				  	})
+				  }
+			  </div>
+			  <div className="bottom">
+				  {
+				  	bottomEntries.map((entry, index) => {
```
Comment (carterbs) [inline src-built-in/components/appLauncher2/src/components/LeftNav.jsx]: There are two schools of thought here, and I don't know where I come down. Maybe @mesner has an opinion.

You can do what you're doing here. Inline array iteration spitting out DOM elements. Alternatively, you can break these out into functions, and just do this: `{this.renderBottomEntries()}`.

I'm not sure which is better but wanted to make you aware of both approaches. I tend to do what you're doing here.

File: src-built-in/components/appLauncher2/src/components/AppDefinition.jsx (hunk 1 above)
Comment (amaroks) [reply src-built-in/components/appLauncher2/src/components/AppDefinition.jsx]: Thanks for pointing to this, did it slightly different 

> {app.tags[index+1] ? `${tag}, ` : `${tag}`}

File: src-built-in/components/appLauncher2/src/components/LeftNav.jsx (hunk 4 above)
Comment (amaroks) [reply src-built-in/components/appLauncher2/src/components/LeftNav.jsx]: I didn't know about the second approach, but I really like it, I have moved the top links to a function and the bottom one to own component, the only reason I didn't break it into more components is that I don't know much about what the store looks like. so I will revisit structure at a later point.

File: src-built-in/components/appLauncher2/src/components/AppDefinition.jsx [hunk 5]
```diff
@@ -13,13 +13,15 @@ export default class AppDefinition extends React.Component {
...
-							app.tags.map((tag, index) => {
-								return <span key={index}>{tag}, </span>
-							})
-						}
-					</div>
+				<div className="app-item-tags">
+				{
+					app.tags.map((tag, index) => {
+						return <span key={index}>
+						{app.tags[index+1] ? `${tag}, ` : `${tag}`}
```
Comment (carterbs) [inline src-built-in/components/appLauncher2/src/components/AppDefinition.jsx]: why +1 here??

File: src-built-in/components/appLauncher2/src/components/Content.jsx [hunk 6]
```diff
@@ -1,37 +1,77 @@
...
+	}
+	
+	onSort(error, data) {
+		this.setState({
+			sortBy: data.value
+		})
+	}
+
+	componentWillMount() {
+		getStore().addListener({field: 'activeFolder'}, this.onActiveFolderChanged)
```
Comment (carterbs) [inline src-built-in/components/appLauncher2/src/components/Content.jsx]: Let's go ahead and use a variable here.

```
let store = getStore();
store.addListener...
```

File: src-built-in/components/appLauncher2/src/components/FilterSort.jsx [hunk 7]
```diff
@@ -1,25 +1,42 @@
...
+			search: event.target.value
+		}, () => {
+			getStore().setValue({
+				field: 'filterText', 
+				value: this.state.search
+			})
+		})
+	}
+
+	onItemClick(item) {
```
Comment (carterbs) [inline src-built-in/components/appLauncher2/src/components/FilterSort.jsx]: Make sure you install some kind of linter. the single quotes will get replaced with double quotes, formatting will fix itself. I highly recommend VSCode.

File: src-built-in/components/appLauncher2/src/components/Content.jsx (hunk 6 above)
Comment (amaroks) [reply src-built-in/components/appLauncher2/src/components/Content.jsx]: Is there a good reason for creating unnecessary variables? Im a believer in "unnecessary variables are bad for your code"

File: src-built-in/components/appLauncher2/src/stores/StoreActions.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
+import {getStore} from './LauncherStore'
+
+export default {
+	getFolders,
```
Comment (carterbs) [inline src-built-in/components/appLauncher2/src/stores/StoreActions.js]: I like this

File: src-built-in/components/appLauncher2/src/components/AppDefinition.jsx (hunk 5 above)
Comment (carterbs) [reply src-built-in/components/appLauncher2/src/components/AppDefinition.jsx]: You are correct. didn't even see the ternary.

File: src-built-in/components/appLauncher2/src/components/Content.jsx (hunk 6 above)
Comment (carterbs) [reply src-built-in/components/appLauncher2/src/components/Content.jsx]: I'm a believer in the rule of 3. If I see an object being used more than 3 times, I'd like to see a variable. For me, variables really help when I'm reading the code. I see `store = getStore()`, and I can file that away mentally.

Then, when I see store later down in the code, I know it's the return value of getStore.

https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming) for a little context

**Pull request closed without merging**